pkg/smartcontract/param_context.go

```diff
@@ -44,219 +26,177 @@ const (
-// valid hex-encoded public keys, Hash256Type for hex-encoded 32 bytes values,
-// SignatureType for hex-encoded 64 bytes values, ByteArrayType for any other
-// valid hex-encoded values and StringType for anything else.
-func inferParamType(val string) ParamType {
-	var err error
-
-	_, err = strconv.Atoi(val)
-	if err == nil {
-		return IntegerType
-	}
-
-	if val == "true" || val == "false" {
-		return BoolType
-	}
+// Parameters is an array of Parameter (TODO: drop it?).
+type Parameters []Parameter
 
-	_, err = address.StringToUint160(val)
-	if err == nil {
-		return Hash160Type
+// TryParseArray converts an array of Parameter into an array of more appropriate things.
+func (p Parameters) TryParseArray(vals ...interface{}) error {
+	var (
+		err error
+		i   int
+		par Parameter
+	)
+	if len(p) != len(vals) {
+		return errors.New("receiver array doesn't fit the Params length")
 	}
-
-	_, err = keys.NewPublicKeyFromString(val)
-	if err == nil {
-		return PublicKeyType
+	for i, par = range p {
+		if err = par.TryParse(vals[i]); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	unhexed, err := hex.DecodeString(val)
-	if err == nil {
-		switch len(unhexed) {
-		case 20:
-			return Hash160Type
-		case 32:
-			return Hash256Type
-		case 64:
-			return SignatureType
+// TryParse converts one Parameter into something more appropriate.
+func (p Parameter) TryParse(dest interface{}) error {
+	var (
+		err  error
+		ok   bool
+		data []byte
+	)
+	switch p.Type {
+	case ByteArrayType:
+		if data, ok = p.Value.([]byte); !ok {
+			return errors.Errorf("failed to cast %s to []byte", p.Value)
+		}
+		switch dest := dest.(type) {
+		case *util.Uint160:
+			if *dest, err = util.Uint160DecodeBytesBE(data); err != nil {
+				return err
+			}
+			return nil
+		case *[]byte:
+			*dest = data
+			return nil
+		case *util.Uint256:
+			if *dest, err = util.Uint256DecodeBytesLE(data); err != nil {
+				return err
+			}
+			return nil
+		case *int64, *int32, *int16, *int8, *int, *uint64, *uint32, *uint16, *uint8, *uint:
+			i := bytesToUint64(data)
+			switch dest := dest.(type) {
+			case *int64:
+				*dest = int64(i)
+			case *int32:
+				*dest = int32(i)
+			case *int16:
+				*dest = int16(i)
+			case *int8:
+				*dest = int8(i)
+			case *int:
+				*dest = int(i)
+			case *uint64:
+				*dest = i
+			case *uint32:
+				*dest = uint32(i)
+			case *uint16:
+				*dest = uint16(i)
+			case *uint8:
+				*dest = uint8(i)
+			case *uint:
+				*dest = uint(i)
+			}
+		case *string:
+			*dest = string(data)
+			return nil
 		default:
-			return ByteArrayType
+			return errors.Errorf("cannot cast param of type %s to type %s", p.Type, dest)
 		}
+	default:
+		return errors.New("cannot define param type")
 	}
-	// Anything can be a string.
-	return StringType
+	return nil
+}
+
+func bytesToUint64(b []byte) uint64 {
+	data := make([]byte, 8)
+	copy(data[8-len(b):], util.ArrayReverse(b))
+	return binary.BigEndian.Uint64(data)
```

Comment:
Though you have to check for length:
https://play.golang.org/p/IgvsG6Nr8iQ
But notice that there actually is a bug in `bytesToUint64` (it will panic on certain inputs)!